providers/azure.go

```diff
@@ -233,13 +234,50 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
+// full claims that can be emitted by AAD are
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens
+// only include email here since we only need email extraction
+type aadIDTokenClaims struct {
+	Email string `json:"email,omitempty"`
+}
+
+// Valid is the function validating the claims,
+// required by jwt.Claims interface.
+// returning nil as we only use it to extract claims instead of validating them
+func (aadIDTokenClaims) Valid() error {
+	return nil
+}
+
+func getEmailFromIDToken(idToken string) (string, error) {
+	claims := &aadIDTokenClaims{}
+	parser := jwt.Parser{}
+
+	// ParseUnverified to extract claims without verifying signature
+	if _, _, err := parser.ParseUnverified(idToken, claims); err != nil {
+		return "", err
+	}
+	if claims.Email == "" {
+		return "", errors.New("missing email claim from id_token")
+	}
+	return claims.Email, nil
+}
+
+// EnrichSessionState finds the email to enrich the session state
+func (p *AzureProvider) EnrichSessionState(ctx context.Context, s *sessions.SessionState) error {
```

Comment:
With your new logic, the return count & conditional branching count in this function is getting quite high. I expect CodeClimate complained and said this needs refactoring?

For instance the profileURL call for JSON & conditional logic for extracting an email from the response is likely best in another private method adjacent to your new `getEmailFromIDToken` implementation.

Reply:
Are you able to get this to an `oidc.IDToken` via an `oidc.Verifier` ([sample](https://github.com/oauth2-proxy/oauth2-proxy/blob/3ff0c23a9e418875cdd7460b0a6a5b40dea71770/providers/oidc_test.go#L151)) and use the `Claims` method like here: https://github.com/oauth2-proxy/oauth2-proxy/blob/3ff0c23a9e418875cdd7460b0a6a5b40dea71770/providers/oidc.go#L233

I figure I'll eventually be on the hook for generic claim -> header support across all providers. When that happens, I'll need to consolidate logic. This would be the 3rd different way to extract IDToken claims, so I'd like to consolidate, at least onto the same `oidc` library if we aren't refactoring the logic into a dedicated method in `utils` at this point.

Reply:
Ah, yep 😉  https://codeclimate.com/github/oauth2-proxy/oauth2-proxy/pull/914

Reply:
i did some refactoring, it should be addressed now.